Package.compiler: add support for opt_flags/debug_flags

scheibelp opened the pull request:
See `mfem` for an example

```
opt_flag_found = any(f in self.compiler.opt_flags for f in cxxflags)
```

this wasn't working, so I added `opt_flags` to the compiler adaptor. `Package.compiler.opt_flags` will return the `opt_flags` for C, C++, *or* Fortran compiler, the first of whichever is defined (in that order). A test is added for this behavior.

`mfem` itself is refactored to use `spec["cxx"].opt_flags`, which is preferred (but the general change above is to support packages in other repos which may not have updated to this form yet).

This also adds a default empty `opt_flags`/`debug_flags` to `CompilerPackage`, so one can ask for `Spec["c"].opt_flags` regardless of what underlying compiler is in use.

~If the package only depends on one language, or uses the same compiler for c/c++, then you can ask for `compiler.opt_flags`; otherwise you ask for `compiler.opt_flags_for_lang(c)`.~

scheibelp committed:
Package.compiler: add support for opt_flags/debug_flags

alalazo commented on lib/spack/spack/compilers/adaptor.py:
```diff
@@ -79,6 +79,31 @@ def implicit_rpaths(self) -> List[str]:
+        compiler_pkg = self._one_compiler("debug_flags").package
+        return getattr(compiler_pkg, "debug_flags", [])
+
+    def opt_flags_for_language(self, lang: str) -> List[str]:
+        return getattr(self.compilers[Languages(lang)].package, "opt_flags", [])
+
+    def debug_flags_for_language(self, lang: str) -> List[str]:
+        return getattr(self.compilers[Languages(lang)].package, "debug_flags", [])
```
The adaptor is a temporary measure to avoid having to change calls to `.compiler.<attribute>` to something like:
```
self[language].<attribute>
```
immediately. The idea is to give users (and our `builtin` repo) a grace period to adapt to having compilers as dependencies. 

I'm fine adding back `debug_flags` / `opt_flags` if there are many packages using them [^1], but I would not add new API that wasn't previously in `Compiler`.

[^1]: Sometimes ago we briefly discussed these options in a Thu meeting, and decided to not put them in the adaptor, due to the fact that less than 10 packages were using them in `builtin`.

scheibelp committed:
update mfem reference

scheibelp committed:
update and test semantics

scheibelp committed:
all compilers have a default (empty) opt_flags and debug_flags

scheibelp committed:
style fixes

scheibelp committed:
rm unused property

scheibelp commented:
Update PR based on discussion (and also updated PR description).

alalazo commented on var/spack/repos/builtin.mock/packages/gcc/package.py:
```diff
@@ -40,6 +40,8 @@ class Gcc(CompilerPackage, Package):
     compiler_version_regex = r"(?<!clang version)\s?([0-9.]+)"
     compiler_version_argument = ("-dumpfullversion", "-dumpversion")
 
+    opt_flags = opt_flags = ["-Otestopt"]
```
```suggestion
    opt_flags = ["-Otestopt"]
```

alalazo commented on lib/spack/spack/test/spec_semantics.py:
```diff
@@ -1365,6 +1365,13 @@ def test_splice_swap_names_mismatch_virtuals(self, default_mock_concretization,
 
+    def test_adaptor_optflags(self):
+        spec = spack.concretize.concretize_one(Spec("pkg-a"))
+        assert "-Otestopt" in spec.package.compiler.opt_flags
+        # This is not set, make sure we get an empty list
+        for x in spec.package.compiler.debug_flags:
+            pass
```
Minor tweaks to try improving clarity of the test:
```suggestion
        compiler_adaptor = spec.package.compiler
        assert compiler_adaptor.opt_flags == ["-Otestopt"]
        assert compiler_adaptor.debug_flags == []
```

alalazo replied:
Another minor improvement could be to check against the mock `GCC.optflags` etc.

scheibelp replied:
I didn't want to test the type of the return value in this test: i.e. I intentionally did not do what you are suggesting. I think this test should be satisfied if the returned value is a sequence.

alalazo commented on lib/spack/spack/test/spec_semantics.py:
```diff
@@ -1365,6 +1365,13 @@ def test_splice_swap_names_mismatch_virtuals(self, default_mock_concretization,
         with pytest.raises(spack.spec.SpliceError, match="virtual"):
             vt.splice(vh, transitive)
 
+    def test_adaptor_optflags(self):
```
```suggestion
    def test_compiler_adaptor_optflags_and_debugflags(self):
        """Tests that we can obtain the list of optflags, and debugflags, from the compiler adaptor, and that this
        list is taken from the appropriate compiler package.
        """
```

alalazo commented on lib/spack/spack/compilers/adaptor.py:
```diff
@@ -79,6 +79,22 @@ def implicit_rpaths(self) -> List[str]:
+
+    @property
+    def debug_flags(self) -> List[str]:
+        return self._first_defined("debug_flags") or []
```
I think here and above you can use:
```suggestion
        return next(iter(self.compilers.values())).package.opt_flags
```
The constructor ensures that there is at least one value, while:

https://github.com/spack/spack/blob/1904d99fd0fc592978cbc8e84321d723a8983a31/lib/spack/spack/compilers/adaptor.py#L202-L213

ensure the dict is constructed with C, followed by C+, followed by Fortran.

alalazo reviewed (commented):
A few minor comments, but essentially LGTM, thanks!

scheibelp committed:
review updates

scheibelp committed:
auto style fix

scheibelp committed:
CompilerPackage always defines property

scheibelp committed:
update docstrings for properties

scheibelp commented:
I followed all requests except for https://github.com/spack/spack/pull/50126#discussion_r2056318690 (where I explain my reasoning).

I added a docstring to the `CompilerAdaptor`, and also updated the docs for `CompilerPackage.{opt_flags,debug_flags}`
